java/ql/src/Security/CWE/CWE-200/TempDirLocalInformationDisclosure.ql

```diff
@@ -102,11 +103,36 @@ private class FileCreateTempFileSink extends FileCreationSink {
   }
 }
 
+/**
+ * A guard that holds when the program is definitely running under some version of Windows.
+ */
+abstract private class WindowsOsBarrierGuard extends DataFlow::BarrierGuard { }
+
+private class IsNotUnixBarrierGuard extends WindowsOsBarrierGuard instanceof IsUnixGuard {
+  override predicate checks(Expr e, boolean branch) {
+    this.controls(e.getBasicBlock(), branch.booleanNot())
+  }
+}
+
+private class IsWindowsBarrierGuard extends WindowsOsBarrierGuard instanceof IsWindowsGuard {
+  override predicate checks(Expr e, boolean branch) { this.controls(e.getBasicBlock(), branch) }
+}
+
+private class IsSpecificWindowsBarrierGuard extends WindowsOsBarrierGuard instanceof IsSpecificWindowsVariant {
+  override predicate checks(Expr e, boolean branch) {
+    branch = true and this.controls(e.getBasicBlock(), branch)
+  }
+}
```

Comment:
These three barrier-guards don't really make sense. Using `Guard.controls` in the `checks` predicate of a barrier-guard is almost guaranteed to be completely wrong, as the barrier-guard abstraction is explicitly designed to handle the `.controls`-aspect of guards.
The expression `e` should be the variable access being checked by the guard, and the `branch` should be the result of the guard that indicates that `e` is safe. So the basic implementation hint for the `checks` predicate should just be to use the AST and nothing more. If you really want to use `Guard.controls` to pick out the nodes that are sanitised then you're really just defining a regular sanitizer and not a sanitizer-guard.

Reply:
So, I partially understand what you're saying. That being said, I don't know how to fix it.

Also, could you suggest a test case that this doesn't work for that I could use as a unit test

Reply:
A barrierguard is just syntactic sugar for a sanitizer with a specific pattern. To see how a barrier-guard gets desugared into a sanitizer, look at the `getAGuardedNode()` predicate:
```
  final Node getAGuardedNode() {
    exists(SsaVariable v, boolean branch, RValue use |
      this.checks(v.getAUse(), branch) and
      use = v.getAUse() and
      this.controls(use.getBasicBlock(), branch) and
      result.asExpr() = use
    )
  }
```
You end up just joining `this.controls` with itself, which semantically is mostly harmless. But you can simplify things by just defining the sanitizers directly. E.g. replace `IsSpecificWindowsBarrierGuard` with the following:
```
abstract private class WindowsOsSanitizer extends DataFlow::Node { }

private class IsSpecificWindowsSanitizer extends WindowsOsSanitizer instanceof IsSpecificWindowsVariant {
  IsSpecificWindowsSanitizer() {
    any(IsSpecificWindowsVariant guard).control(this.asExpr().getBasicBlock(), true)
  }
}

...

 override predicate isSanitizer(DataFlow::Node node) {
    node instanceof WindowsOsSanitizer
  }
```
This is mostly equivalent to your `IsSpecificWindowsBarrierGuard` except it drops the restriction that only uses of ssa variables are sanitized.
The other two, `IsNotUnixBarrierGuard` and `IsWindowsBarrierGuard`, can be similarly rewritten, but note that they are somewhat broken as-is, since they currently guard both of their branches. In
```
if (<IsWindowsBarrierGuard>) {
  foo
} else {
  bar
}
```
you're sanitizing both `foo` and `bar`.